### PR TITLE
Delegation registry integration with merkle-proof minter

### PR DESCRIPTION
**Pull request by jakerockland**
Integration the delegate.cash delegation registry with our merkleminter – co-authored with @mchrupcala, who added the integration testing support here.

Diff between this version (V2) and the V1 merkle minter that it branches from: https://www.diffchecker.com/nZgT92ry

Additional context: 
- https://0xfoobar.substack.com/p/delegatecash 
- https://twitter.com/0xfoobar/status/1575290861570740226?s=20&t=llZHT6mXd_RDYB6CrSrgpw 
- https://github.com/0xfoobar/delegation-registry

### CODE REVIEW

**Comment by jakerockland** [comment]
cc @yoshiwarab

File: contracts/minter-suite/Minters/MinterMerkle/MinterMerkleV1.sol
Comment (ryley-o) [inline contracts/minter-suite/Minters/MinterMerkle/MinterMerkleV1.sol:329]: I think minting to `msg.sender` is okay here, but we should make this behavior very clear on the frontend, because I can definitely see some users thinking the mint will be sent directly to their vault. We might even want to only accept vault to be specified on calls to `purchaseTo` (and assume delegee of zero here).
Honestly, this whole concept of "should a delegate claiming an airdrop/benefit of any kind be the receiver of said benefit" might be a good "best practice" question for the registry dev chat.

**Review by ryley-o — COMMENTED** [review]
This approach generally LGTM! Very straightforward 😄 

Couple of notes:
- You will have some merge conflicts after #321 is merged
- Our super cool `purchaseTo_K1L` will have to be be updated 😓

File: contracts/minter-suite/Minters/MinterMerkle/MinterMerkleV1.sol
Comment (mchrupcala) [reply contracts/minter-suite/Minters/MinterMerkle/MinterMerkleV1.sol:327]: +1 I think it's easy to mix up delegee vs. delegate 😅

File: contracts/minter-suite/Minters/MinterMerkle/MinterMerkleV1.sol
Comment (mchrupcala) [inline contracts/minter-suite/Minters/MinterMerkle/MinterMerkleV1.sol:433]: Is this error message something that we'll surface to the end-user? Maybe something like "No match was found for this delegate/vault pair".

File: contracts/minter-suite/Minters/MinterMerkle/MinterMerkleV1.sol
Comment (mchrupcala) [reply contracts/minter-suite/Minters/MinterMerkle/MinterMerkleV1.sol:329]: The funds for minting are still coming from the delegate, right? If so then I agree it feels to me like our existing purchaseTo() and the delegate.cash purchasing are very similar user flows (difference being the delegation would automatically pull the registered vault address for the user instead of a copy/paste). As such I'm not sure if I see the use case where purchase() would need a `delegee` because implicitly this flow is always meant for sending $$ & receiving art to/from the same wallet.

^This has me wondering if we want to re-work the purchase modal along the lines of "purchase" and then a new screen or section that asks "are you sending this to another wallet?" --> if yes, then a copy/paste uses `purchaseTo()` or if getDelegatesForContract returns 1+ addresses, the delegate sees a list of vaults they can click & send to.

**Comment by jakerockland** [comment]
> You will have some merge conflicts after #321 is merged

Yes yes! Was planning to wait until #321 is landed to do another real pass over updating this PR, and to add test coverage, etc.

> Our super cool purchaseTo_K1L will have to be be updated 😓

Indeed 🥲 I think I will need your assistance (or direction towards the tooling to use) to determine most efficient signature to use here ser.

File: contracts/minter-suite/Minters/MinterMerkle/MinterMerkleV1.sol
Comment (jakerockland) [reply contracts/minter-suite/Minters/MinterMerkle/MinterMerkleV1.sol:433]: Generally would say that we probably should not design our contract assertion messages to be as detailed/verbose as error messages should be for user-facing frontend applications, as there are reasons to keep bytes used for assertions messages smol-er. 

In the case of surfacing explanation of this type of revert scenario to users, would say that we should catch the error and output more detailed error message as frontend handling itself. :)

File: contracts/minter-suite/Minters/MinterMerkle/MinterMerkleV1.sol
Comment (jakerockland) [reply contracts/minter-suite/Minters/MinterMerkle/MinterMerkleV1.sol:13]: Agreed–will bump to 0.8.16 when I do another edit-pass over this PR (leaving this comment unresolved until I address), which I'll do after #321 is merged 👍 💜

File: contracts/minter-suite/Minters/MinterMerkle/MinterMerkleV1.sol
Comment (jakerockland) [reply contracts/minter-suite/Minters/MinterMerkle/MinterMerkleV1.sol:327]: SGTM–will update to rename when I do another edit-pass over this PR (leaving this comment unresolved until I address), which I'll do after #321 is merged 👍 💜

File: contracts/minter-suite/Minters/MinterMerkle/MinterMerkleV1.sol
Comment (jakerockland) [reply contracts/minter-suite/Minters/MinterMerkle/MinterMerkleV1.sol:329]: Yeah this is a great point–wonder in this case if we may even want to just disallow `purchase` to be used when using this minter via the `delegate-vault` approach, instead _only allowing_ `purchaseTo`, just to simplify the set of possible interactions with this minter to be **only 3 options**:

1) minting as a directly-allowlisted minter (on the merkle list directly) using `purchase` (`msg.sender` receives NFT)
2) minting as a directly-allowlisted minter (on the merkle list directly) using `purchaseTo` (provided `_to` address receives NFT)
3) minting as the delegate for an allowlisted minter (vault is on merkle list, minter is not but is a delegate) using `purchaseTo` (provided `_to` address receives NFT)

>  Honestly, this whole concept of "should a delegate claiming an airdrop/benefit of any kind be the receiver of said benefit" might be a good "best practice" question for the registry dev chat.

This is a great point–going to post this convo there for any additional thoughts, and will also let this bop around in my brain a bit more over the weekend before I loop back and update this PR at all to account for this discussion and the others mentioned here.

**Comment by ryley-o** [comment]
>Indeed 🥲 I think I will need your assistance (or direction towards the tooling to use) to determine most efficient signature to use here ser.

happy to help, feel free to play around at: https://emn178.github.io/solidity-optimize-name/

(I always double check the signature in remix though because that frontend allows some entry of invalid solidity 🤷 e.g. use `uint256` not `uint`, but that frontend incorrectly allows `uint`)

File: contracts/minter-suite/Minters/MinterMerkle/MinterMerkleV1.sol
Comment (jakerockland) [reply contracts/minter-suite/Minters/MinterMerkle/MinterMerkleV1.sol:13]: Resolving now that we're on 0.8.17 anyway 😄

File: contracts/minter-suite/Minters/MinterMerkle/MinterMerkleV1.sol
Comment (jakerockland) [reply contracts/minter-suite/Minters/MinterMerkle/MinterMerkleV1.sol:327]: Updated.

File: contracts/minter-suite/Minters/MinterMerkle/MinterMerkleV1.sol
Comment (jakerockland) [reply contracts/minter-suite/Minters/MinterMerkle/MinterMerkleV1.sol:329]: Updating to just support these 3 routes: `purchase`, `purchaseTo`, and `purchaseTo` as a delegate–I think this is the most explicit pattern and allows flexibility on the frontend for us to iterate on UX to make it as clear as possible.

**Comment by jakerockland** [comment]
> > Indeed 🥲 I think I will need your assistance (or direction towards the tooling to use) to determine most efficient signature to use here ser.
> 
> happy to help, feel free to play around at: https://emn178.github.io/solidity-optimize-name/
> 
> (I always double check the signature in remix though because that frontend allows some entry of invalid solidity 🤷 e.g. use `uint256` not `uint`, but that frontend incorrectly allows `uint`)

Huzzah! Thank you for the tip! Updating to `purchaseTo_kem` based on tweakin w/ that tool 🙌 💜

**Comment by jakerockland** [comment]
@mchrupcala @ryley-o still have not yet had a chance to update to add tests, but did take a pass over incorporating the other feedback in this PR–if I have any extra time during Devcon week next week will try to get test coverage added so that we can get this in a more merge-ready state for real review here.

**Comment by jakerockland** [comment]
Also, if either of you would like to take on getting this across the finish line in that regard, please do not block on me–that said, please also do not let this derail other priorities that are non-Merkle related, this is a stretch goal after all 💜 😄

**Comment by jakerockland** [comment]
@mchrupcala as part of taking this PR on, I would recommend that we bump the version on the MerkleMinter from V1 to V2, given that we've now deployed something to mainnet with the V1 nomenclature that does not include this functionality. 

cc @ryley-o for his thoughts as well to see if there is a difference of opinion from his perspective.

**Comment by jakerockland** [comment]
Additional context re: MerkleMinterV1 being live on mainnet here: https://github.com/ArtBlocks/artblocks-contracts/pull/337

🙏 💜

File: test/minter-suite-minters/MinterMerkle/MinterMerkleV1.test.ts
Comment (mchrupcala) [inline test/minter-suite-minters/MinterMerkle/MinterMerkleV1.test.ts:176]: @ryley-o found a great pattern using [Smock](https://smock.readthedocs.io/en/latest/fakes.html#options) so that we can mock `checkDelegateForContract`. In the future we can prob use this to mock method calls in our unit tests whenever we have an external contract dependency.

File: package.json
Comment (mchrupcala) [inline package.json:55]: Wanted to call out the Hardhat version upgrade - I had to switch versions to fix a typeerror that Smock was giving me (`Cannot read properties of undefined (reading 'events')`) but please lmk if we wanted to stay consistent with the earlier version 🙏🏻

File: test/libs/BytecodeStorage_BytecodeTextCR_DMock.test.ts
Comment (mchrupcala) [inline test/libs/BytecodeStorage_BytecodeTextCR_DMock.test.ts:319]: One more callout- `expectRevert.unspecified()` stopped working for me (I'm guessing that might be due to the Hardhat upgrade, although this util's from OpenZeppelin). Adding a revert error seemed to do the trick!

**Comment by mchrupcala** [comment]
> @mchrupcala as part of taking this PR on, I would recommend that we bump the version on the MerkleMinter from V1 to V2, given that we've now deployed something to mainnet with the V1 nomenclature that does not include this functionality.
> 
> cc @ryley-o for his thoughts as well to see if there is a difference of opinion from his perspective.

Bumped! I switched this to `MinterMerkleV2` everywhere I could find a reference.

File: hardhat.config.ts
Comment (jakerockland) [inline hardhat.config.ts:65]: I don't think this is needed.

File: README.md
Comment (jakerockland) [inline README.md:87]: This diff seems off to me–could you clarify please?

File: package.json
Comment (jakerockland) [reply package.json:55]: This SGTM 🙏

File: README.md
Comment (mchrupcala) [reply README.md:87]: Oh sorry, I was assuming since we haven't deployed core V3 to prod then our V1 minter should be replaced by V2, instead of listing both in the ReadMe. Is it better if I list both V1 and V2 instead of replacing this one?

File: hardhat.config.ts
Comment (jakerockland) [reply hardhat.config.ts]: (reason being that this is already duplicated above)

File: test/minter-suite-minters/MinterMerkle/MinterMerkleV1.test.ts
Comment (jakerockland) [reply test/minter-suite-minters/MinterMerkle/MinterMerkleV1.test.ts]: 🙌

File: hardhat.config.ts
Comment (jakerockland) [reply hardhat.config.ts]: Removed.

File: test/minter-suite-minters/MinterMerkle/MinterMerkleV2.test.ts
Comment (jakerockland) [inline test/minter-suite-minters/MinterMerkle/MinterMerkleV2.test.ts:181]: @mchrupcala rather than adding this as a shared before-each assumption, can we please just set this assumption on a per-test level to make things more explicit / more in alignment with assumptions of each individual test? I think this will improve read-ability.

**Review by jakerockland — COMMENTED** [review]
One outstanding item @mchrupcala and then I think we're g2g re: merging this and beginning to follow up on the minor frontend pieces here.

File: test/minter-suite-minters/MinterMerkle/MinterMerkleV2.test.ts
Comment (mchrupcala) [reply test/minter-suite-minters/MinterMerkle/MinterMerkleV2.test.ts]: Agreed! And this way the next time somebody writes a delegation test they won't inherit the `.returns(true)`. One more note- I added a comment to mention the deployed address we're mocking for the delegation contract.

Would you like me to go ahead and merge this? I'll request another round of review from Ryley in the meantime for when he's back.

**Review by mchrupcala — APPROVED** [review]

File: test/minter-suite-minters/MinterMerkle/MinterMerkleV2.test.ts
Comment (jakerockland) [reply test/minter-suite-minters/MinterMerkle/MinterMerkleV2.test.ts]: Fantastic–yep this LGTM to merge. 🙏

**Comment by jakerockland** [comment]
Merging this @mchrupcala – I think we're ready to take on frontend changes necessary for this and to test this end-to-end on Goerli once that is done 🔥 💜 

Great work here!!

**Pull request merged**
